ros/src/waypoint_updater/jmt.py

```diff
@@ -0,0 +1,170 @@
+
+    # def JMTD_at(self, displacement, coeffs, t0, tmax, deq_wpt_ptr):
+    def JMTD_at(self, displacement, t0, tmax):
+        # find JMT descriptors at displacement
+        s_last = 0.0
+        t_found = False
+        t_inc = 0.01
+        iterations = 0
+
+        for t_cnt in range(int((tmax-t0)/t_inc) + 100):
+            iterations += 1
+            t = t0 + t_cnt * t_inc
+            s = self.get_s_at(t)
+            if s > displacement:
+                t = t - (1 - (s - displacement) / (s - s_last)) * t_inc
+                t_found = True
+                break
+            # end if
+            s_last = s
+        # end for
+        if t_found is False:
+            rospy.loginfo("waypoint_updater:JMTD_at Ran out of bounds without "
+                          "finding target displacement")
+            return None
+
+        s = self.get_s_at(t)
+        delta_s = (displacement - s)
+        if delta_s > 0.0:
+            searchdir = 1.0
+        else:
+            searchdir = -1.0
+
+        t_smallinc = searchdir * 0.005
+        while delta_s * searchdir > 0.0:
+            iterations += 1
+            t += t_smallinc
+            delta_s = displacement - self.get_s_at(t)
+
+        rospy.loginfo("delta_s = {}, t= {}, iterations = {}".format(
+            delta_s, t, iterations))
+        if delta_s > 0.1:
+            rospy.loginfo("waypoint_updater:JMTD_at need to refine algo,"
+                          " delta_s is {}".format(delta_s))
+
+        details = JMTDetails(self.get_s_at(t), self.get_v_at(t),
+                             self.get_a_at(t), self.get_j_at(t), t)
+
+        rospy.loginfo("waypoint_updater:JMTD_at displacement {} found "
+                      "s,v,a,j,t = {}".format(displacement, details))
+
+        return details
+
+    def get_s_at(self, t):
+        return self.coeffs[0] + self.coeffs[1] * t + self.coeffs[2] * t**2 +\
+            self.coeffs[3] * t**3 + self.coeffs[4] * t**4 + self.coeffs[5] *\
+            t**5
+
+    def get_v_at(self, t):
+        return self.coeffs[1] + 2.0 * self.coeffs[2]*t + 3.0 *\
+            self.coeffs[3] * t**2 + 4.0 * self.coeffs[4] *\
+            t**3 + 5.0 * self.coeffs[5] * t**4
+
+    def get_a_at(self, t):
+        return 2.0 * self.coeffs[2] + 6.0 * self.coeffs[3] * t + 12.0 *\
+            self.coeffs[4] * t**2 + 20.0 * self.coeffs[5] * t**3
+
+    def get_j_at(self, t):
+        return 6.0 * self.coeffs[3] + 24.0 * self.coeffs[4] * t + 60.0 *\
+            self.coeffs[5] * t**2
+
+
+class JMTDetails(object):
+    def __init__(self, S, V, A, J, t):
+        self.S = S
+        self.V = V
+        self.A = A
+        self.J = J
+        self.time = t
+
+    def set_VAJt(self, V, A, J, time):
+        self.V = V
+        self.A = A
+        self.J = J
+        self.time = time
+
+    def __repr__(self):
+        return "%5.3f, %2.4f, %2.4f, %2.4f, %2.3f" % (self.S, self.V, self.A,
+                                                      self.J, self.time)
```

Comment:
For easier plotting in Excel charts, can time be moved to the first position before S?  This way after pasting the CSV into Excel, the columns can be highlighted in a block and it picks up time as the X-axis.

Reply:
will change.  I was moving the columns around myself in spreadsheets.